File: backend/src/main/java/com/crm/entity/Benutzer.java
package com.crm.entity;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

import com.crm.entity.enums.BenutzerRolle;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;

@Entity
@Table(name = "benutzer")
public class Benutzer {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, length = 50)
    private String benutzername;

    @Column(nullable = false)
    private String passwort;

    @Column(nullable = false, length = 100)
    private String vorname;

    @Column(nullable = false, length = 100)
    private String nachname;

    @Column(nullable = false, unique = true, length = 255)
    private String email;

    @Column(nullable = false)
    private boolean aktiv = true;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "benutzer_rollen", joinColumns = @JoinColumn(name = "benutzer_id"))
    @Enumerated(EnumType.STRING)
    @Column(name = "rolle")
    private Set<BenutzerRolle> rollen = new HashSet<>();

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getBenutzername() { return benutzername; }
    public void setBenutzername(String benutzername) { this.benutzername = benutzername; }
    public String getPasswort() { return passwort; }
    public void setPasswort(String passwort) { this.passwort = passwort; }
    public String getVorname() { return vorname; }
    public void setVorname(String vorname) { this.vorname = vorname; }
    public String getNachname() { return nachname; }
    public void setNachname(String nachname) { this.nachname = nachname; }
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    public boolean isAktiv() { return aktiv; }
    public void setAktiv(boolean aktiv) { this.aktiv = aktiv; }
    public Set<BenutzerRolle> getRollen() { return rollen; }
    public void setRollen(Set<BenutzerRolle> rollen) { this.rollen = rollen; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
}

File: ciam/src/main/kotlin/com/crm/ciam/entity/Benutzer.kt
package com.crm.ciam.entity

import com.crm.ciam.entity.enums.BenutzerRolle
import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "benutzer")
class Benutzer {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null

    @Column(nullable = false, unique = true, length = 50)
    var benutzername: String = ""

    @Column(nullable = false)
    var passwort: String = ""

    @Column(nullable = false, length = 100)
    var vorname: String = ""

    @Column(nullable = false, length = 100)
    var nachname: String = ""

    @Column(nullable = false, unique = true, length = 255)
    var email: String = ""

    @Column(nullable = false)
    var aktiv: Boolean = true

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "benutzer_rollen", joinColumns = [JoinColumn(name = "benutzer_id")])
    @Enumerated(EnumType.STRING)
    @Column(name = "rolle")
    var rollen: MutableSet<BenutzerRolle> = mutableSetOf()

    @Column(nullable = false, updatable = false)
    var createdAt: LocalDateTime? = null

    @Column(nullable = false)
    var updatedAt: LocalDateTime? = null

    @PrePersist
    protected fun onCreate() {
        createdAt = LocalDateTime.now()
        updatedAt = LocalDateTime.now()
    }

    @PreUpdate
    protected fun onUpdate() {
        updatedAt = LocalDateTime.now()
    }
}
